### PR TITLE
fix: "weight_norm_fwd_first_dim_kernel" not implemented for 'BFloat16'

### LINKED ISSUE #107914: [BUG] "weight_norm_fwd_first_dim_kernel" not implemented for 'BFloat16'
### 🐛 Describe the bug

Attempting to `weight_norm` a bfloat16 layer in a CUDA environment results in an error like 
```RuntimeError: "weight_norm_fwd_first_dim_kernel" not implemented for 'BFloat16'```

No error on `weight_norm` in CPU environment, but only in CUDA environment.

# errror
```
"weight_norm_fwd_first_dim_kernel" not implemented for 'BFloat16'
  File "[user-dir]/lib/python3.10/site-packages/torch/nn/utils/weight_norm.py", line 25, in compute_weight
    return _weight_norm(v, g, self.dim)
  File "[user-dir]/lib/python3.10/site-packages/torch/nn/utils/weight_norm.py", line 50, in apply
    setattr(module, name, fn.compute_weight(module))
  File "[user-dir]/lib/python3.10/site-packages/torch/nn/utils/weight_norm.py", line 109, in weight_norm
    WeightNorm.apply(module, name, dim)
  File "/home/jsb193/workspace/github/llm/LLM42/train/train_dpo.py", line 48, in <module>
    nn.utils.weight_norm(module, "weight")
RuntimeError: "weight_norm_fwd_first_dim_kernel" not implemented for 'BFloat16'
```

# example
```python
import torch.nn as nn

module = torch.nn.Linear(20, 40)
module = module.to(torch.bfloat16)
module = module.to("cuda")
nn.utils.weight_norm(module, "weight")
```

### Versions

pytorch: 2.0.1+cu117
python: 3.10.11
os: ubuntu-18.04
gpu: GeForce3090

### PR DESCRIPTION
**Pull request by jp1924**
Fixes #107914

### CODE REVIEW

**Comment by jp1924** [comment]
@mikaylagawarecki 
I'm not very good with github, so I didn't set up a reviewer, could you review it?

**Comment by jp1924** [comment]
Then I looked at dispatch and it doesn't support both bfloat16 and float32, 16, and half, so I created `AT_DISPATCH_ALL_FLOATING_TYPES`, is that appropriate?

**Review by mikaylagawarecki — COMMENTED** [review]
I believe you can use `AT_DISPATCH_FLOATING_TYPES_AND2`, so you don't need to define a new dispatch macro :)

https://github.com/pytorch/pytorch/blob/691de5e4ec359619f8ba87baaf78779ae0959a2d/aten/src/ATen/Dispatch.h#L279

e.g. `AT_DISPATCH_FLOATING_TYPES_AND2(kBFloat16, kHalf, ...` like here 

https://github.com/pytorch/pytorch/blob/fe1f26af8a13de66bbd60183d3b8c679e1b34e26/aten/src/ATen/native/cuda/Normalization.cu#L97

Could you also update [this test](https://github.com/pytorch/pytorch/blob/main/test/test_nn.py#L1735) to make sure bfloat16 works as expected on CUDA.

As a bonus, it would be nice if you could add an opinfo for `torch._weight_norm` (but that is not necessary for this PR :). Let me know if you would like more details there!

File: test/test_nn.py [hunk 1]
```diff
@@ -1772,6 +1772,12 @@ def test_weight_norm(self):
         m = nn.Linear(4, 5, dtype=torch.float16)
         m = torch.nn.utils.weight_norm(m)
 
+        m = nn.Linear(4, 5, dtype=torch.bfloat16, device="cpu")
```
Comment (mikaylagawarecki) [inline test/test_nn.py:1775]: hm I think instead we should move `test_weight_norm` to  this class https://github.com/pytorch/pytorch/blob/2daf300b5e4489263261b4b943c5aca4a4e1da9c/test/test_nn.py#L7828

The implication here is that the test will be parametrized over device, and then we can pass device=device when we create the `nn.Linear`, you can look at other examples within the class for inspiration.

File: aten/src/ATen/native/cuda/WeightNorm.cu [hunk 2]
```diff
@@ -453,8 +453,8 @@ std::tuple<Tensor, Tensor> weight_norm_backward_cuda
 
     cudaStream_t stream = at::cuda::getCurrentCUDAStream();
 
-    AT_DISPATCH_FLOATING_TYPES_AND_HALF
-      (saved_v.scalar_type(),
+    AT_DISPATCH_FLOATING_TYPES_AND2
+      (kBFloat16, kHalf, v.scalar_type(),
```
Comment (mikaylagawarecki) [inline aten/src/ATen/native/cuda/WeightNorm.cu:457]:
```suggestion
      (kBFloat16, kHalf, saved_v.scalar_type(),
```

File: aten/src/ATen/native/cuda/WeightNorm.cu [hunk 3]
```diff
@@ -486,8 +486,8 @@ std::tuple<Tensor, Tensor> weight_norm_backward_cuda
...
```
Comment (mikaylagawarecki) [inline aten/src/ATen/native/cuda/WeightNorm.cu:490]:
```suggestion
      (kBFloat16, kHalf, saved_v.scalar_type(),
```

**Review by mikaylagawarecki — COMMENTED** [review]
three things

1) the remaining CI failure is because the following line needs to be updated to account for BFloat16 (under the if-condition with half)
https://github.com/pytorch/pytorch/blob/eea7d3b589c5a9100593769da0c116ac0f413231/aten/src/ATen/native/cuda/WeightNorm.cu#L350-L351
An even better update here for modularity might be to use `at::opmath_type<T>` here [like such](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/group_norm_kernel.cpp#L56). Let me know if you want more context on this


2) Separately, it occurred to me while looking into this that we have [recently deprecated `torch.nn.utils.weight_norm`](https://github.com/pytorch/pytorch/blob/1cd8a97679f63bbf3efc64cbea13c63fe80c7bff/torch/nn/utils/weight_norm.py#L30) as a user facing API in favor of `torch.nn.utils.parametrizations.weight_norm`, (which calls into the same CUDA kernel, so this fix is still valid!) so we should add/update the testing for  `torch.nn.utils.parametrizations.weight_norm` instead

3) Apologies that it has been a long back and forth with this PR. Let me know if you have questions about building locally and running tests locally, I would be most happy to help 😄

**Comment by jp1924** [comment]
@mikaylagawarecki  Thanks for the kind words!

I've been fiddling with it too, as I was getting stuck on compiling pytorch, but I figured it out and am testing it out now.

# about error
I agree with the code modification you suggested.     
      

https://github.com/pytorch/pytorch/blob/eea7d3b589c5a9100593769da0c116ac0f413231/aten/src/ATen/native/cuda/WeightNorm.cu#L380
      
However, upon further debugging, I found that there is an issue with this part as well. (The error occurs even if `AccType` is BFloat16)
      
https://github.com/pytorch/pytorch/blob/eea7d3b589c5a9100593769da0c116ac0f413231/aten/src/ATen/templates/TensorMethods.cpp#L11-L17
I'm not skilled in C++, but 
When the code uses `norms.mutable_data_ptr<accscalar_t>()`, the `ScalarType type` is set to `Float`, and I encountered the error `message expected scalar type float but found BFloat16`.
      
However, I checked that `norms.mutable_data_ptr` works normally when used with `norms.mutable_data_ptr<scalar_t>()` or `norms.mutable_data_ptr<at::BFloat16>()`.
     
I think it's necessary to take a closer look at `accscalar_t`.

# bug...?

I have trouble understanding a certain part.      
     
When I use `torch.bfloat16`, it follows the path `WeightNorm.cu > weight_norm_cuda`, but when I use `torch.float16`, it seems to take a different code path and does not go through `weight_norm_cuda`.
      
I tested this using the code below, and for bfloat16, it displayed 'is cuda weight_norm!', but for float16, it didn't display anything.
```python
# bfloat16
import torch; torch.nn.utils.weight_norm(torch.nn.Linear(1,2, dtype=torch.bfloat16, device="cuda"), "weight")
# float16
import torch; torch.nn.utils.weight_norm(torch.nn.Linear(1,2, dtype=torch.float16, device="cuda"), "weight")
```
      
My thought was that I could modify `Bfloat16` like float16 since it worked fine with `float16`, but I don't know what to do with `bfloat16` since `float16` is going somewhere else.
```c++
    scalar_t* x = norms.mutable_data_ptr<scalar_t>();
    accscalar_t* y = (accscalar_t*) x;
```
I tried to make it work by TypeCasting, but I don't think that's the right way to do it.

# my test code
The above problem was tested using the following code
      
```c++
// if you want to test, please input this includes
#include <c10/util/BFloat16.h>
#include <typeinfo>
#include <ATen/OpMathType.h>
#include <iostream>
...
std::tuple<Tensor,Tensor> weight_norm_cuda
  (const Tensor & v,
   const Tensor & g,
   int64_t dim)
{
  std::cout << "is cuda weight_norm!" << std::endl;
  auto w = at::empty_like(v, LEGACY_CONTIGUOUS_MEMORY_FORMAT);

  // weight_norm_fused does have a derivative defined in derivatives.yaml, therefore, VariableType.cpp
  // sends the unpacked g.data() as the argument.  In other words, we expect "g" is a bare Tensor here.

  // norms is only needed to stash for backward.
  // g.scalar_type() may be at::ScalarType::Double, Float, or Half.
  // If Half, stash norms as float.
  
  //// NOTE: original code
  // at::ScalarType AccType = g.scalar_type() == at::ScalarType::Half ?
  //                          at::ScalarType::Float : g.scalar_type();

  at::ScalarType AccType = g.scalar_type() == at::ScalarType::BFloat16
      ? at::ScalarType::BFloat16
      : g.scalar_type();

  // auto norms = at::empty_strided(g.sizes(), g.strides(), g.options().dtype(at::BFloat16));
  auto norms = at::empty_strided(g.sizes(), g.strides(), g.options().dtype(AccType));


  const int ndims = v.dim();

  if(dim == 0)
  {
    // Find logical size of each flattened slowest-dim row
    int rowSize = 1;
    for(int i = ndims - 1; i > 0; i--)
      rowSize *= v.size(i);

    cudaStream_t stream = at::cuda::getCurrentCUDAStream();

    printf("____________test_1____________\n");
    AT_DISPATCH_FLOATING_TYPES_AND2(
        kHalf,
        kBFloat16,
        v.scalar_type(), "weight_norm_fwd_first_dim_kernel", [&] {
          using accscalar_t = acc_type<scalar_t, true>;
          using opmath_t = at::opmath_type<scalar_t>;

          std::cout << "test code section norms.mutable_data_ptr in!" << std::endl;
          std::cout << typeid(scalar_t).name() << std::endl; // check scalar_t type
          std::cout << typeid(accscalar_t).name() << std::endl; // check accscalar_t type
          std::cout << typeid(opmath_t).name() << std::endl; // check opmath_t type

          std::cout << norms.dtype() << std::endl; // check norms dtype

          scalar_t* x = norms.mutable_data_ptr<scalar_t>();
          accscalar_t* y = (accscalar_t*) x;

          std::cout << typeid(y).name() << std::endl; // check TypeCasted obj dtype
          std::cout << "test code section norms.mutable_data_ptr out!" << std::endl;

          weight_norm_fwd_first_dim_kernel<scalar_t, accscalar_t>
              <<<v.size(0), BLOCK, BLOCK * sizeof(accscalar_t), stream>>>(
                  w.mutable_data_ptr<scalar_t>(),
                  y, // tmp, it just for testing
                  v.const_data_ptr<scalar_t>(),
                  g.const_data_ptr<scalar_t>(),
                  rowSize);
          C10_CUDA_KERNEL_LAUNCH_CHECK();
        });
  }
  else if(dim == ndims - 1)
  {
    // Precompute slower_dims_size and fast_dim_size
    int slower_dims_size = 1;
    for(int i = 0; i < ndims - 1; i++)
      slower_dims_size *= v.size(i);

    int fast_dim_size = v.size(ndims-1);

    cudaStream_t stream = at::cuda::getCurrentCUDAStream();

    printf("____________test_2____________\n");
    AT_DISPATCH_FLOATING_TYPES_AND2(
        kHalf,
        kBFloat16,
        v.scalar_type(),
        "weight_norm_fwd_last_dim_kernel",
        [&] {
          using accscalar_t = acc_type<scalar_t, true>;
          std::cout << typeid(accscalar_t).name() << std::endl;

          weight_norm_fwd_last_dim_kernel<scalar_t, accscalar_t>
              <<<(fast_dim_size + TILE_W - 1) / TILE_W,
                 dim3(TILE_W, TILE_H),
                 (TILE_W * TILE_H + TILE_W) * sizeof(accscalar_t),
                 stream>>>(
                  w.mutable_data_ptr<scalar_t>(),
                  norms.mutable_data_ptr<accscalar_t>(),
                  v.const_data_ptr<scalar_t>(),
                  g.const_data_ptr<scalar_t>(),
                  fast_dim_size,
                  slower_dims_size);
          C10_CUDA_KERNEL_LAUNCH_CHECK();
        });
  }

  // The kernel execution is asynchronous, so this will only catch errors on the kernel launch,
  // not the kernel's execution.  Errors in kernel execution aren't guaranteed to be caught
  // until a later error check on a synchronizing CUDA call.  Unfortunately, without manually
  // synchronizing here, the foregoing is the best we can do.

  return std::tuple<Tensor, Tensor>{w, norms};
}
  }
```

**Comment by jp1924** [comment]
@mikaylagawarecki  Thanks for the kind words!

I've been fiddling with it too, as I was getting stuck on compiling pytorch, but I figured it out and am testing it out now.

# about error
https://github.com/pytorch/pytorch/blob/eea7d3b589c5a9100593769da0c116ac0f413231/aten/src/ATen/native/cuda/WeightNorm.cu#L380
upon further debugging, I found that there is an issue with this part as well. (The error occurs even if `AccType` is BFloat16)
      
https://github.com/pytorch/pytorch/blob/eea7d3b589c5a9100593769da0c116ac0f413231/aten/src/ATen/templates/TensorMethods.cpp#L11-L17
I'm not skilled in C++, but 
When the code uses `norms.mutable_data_ptr<accscalar_t>()`(or `norms.mutable_data_ptr<opmath_t>()`), the `ScalarType type` is set to `Float`, and I encountered the error `message expected scalar type float but found BFloat16`.
      
However, I checked that `norms.mutable_data_ptr` works normally when used with `norms.mutable_data_ptr<scalar_t>()` or `norms.mutable_data_ptr<at::BFloat16>()`.
     
I think it's necessary to take a closer look at `accscalar_t`.

```c++
    scalar_t* x = norms.mutable_data_ptr<scalar_t>();
    accscalar_t* y = (accscalar_t*) x;
```
now I tried to make it work by TypeCasting, but I don't think that's the right way to do it.

# bug...?

I have trouble understanding a certain part.      
     
When I use `torch.bfloat16`, it follows the path `WeightNorm.cu > weight_norm_cuda`, but when I use `torch.float16`, it seems to take a different code path and does not go through `weight_norm_cuda`.
      
I tested this using the code below, and for bfloat16, it displayed `is cuda weight_norm!`, but for float16, it didn't display anything.
```python
# bfloat16
import torch; torch.nn.utils.weight_norm(torch.nn.Linear(1,2, dtype=torch.bfloat16, device="cuda"), "weight")
# float16
import torch; torch.nn.utils.weight_norm(torch.nn.Linear(1,2, dtype=torch.float16, device="cuda"), "weight")
```
      
My thought was that I could modify `Bfloat16` like float16 since it worked fine with `float16`, but I don't know what to do with `bfloat16` since `float16` is going somewhere else.

# my test code
The above problem was tested using the following code
I'm newbie to C++, so I'm not very good at it yet, sorry.   😅
     
```c++
// if you want to test, please input this includes
#include <c10/util/BFloat16.h>
#include <typeinfo>
#include <ATen/OpMathType.h>
#include <iostream>
...
std::tuple<Tensor,Tensor> weight_norm_cuda
  (const Tensor & v,
   const Tensor & g,
   int64_t dim)
{
  std::cout << "is cuda weight_norm!" << std::endl;
  auto w = at::empty_like(v, LEGACY_CONTIGUOUS_MEMORY_FORMAT);

  // weight_norm_fused does have a derivative defined in derivatives.yaml, therefore, VariableType.cpp
  // sends the unpacked g.data() as the argument.  In other words, we expect "g" is a bare Tensor here.

  // norms is only needed to stash for backward.
  // g.scalar_type() may be at::ScalarType::Double, Float, or Half.
  // If Half, stash norms as float.
  
  //// NOTE: original code
  // at::ScalarType AccType = g.scalar_type() == at::ScalarType::Half ?
  //                          at::ScalarType::Float : g.scalar_type();

  at::ScalarType AccType = g.scalar_type() == at::ScalarType::BFloat16
      ? at::ScalarType::BFloat16
      : g.scalar_type();

  // auto norms = at::empty_strided(g.sizes(), g.strides(), g.options().dtype(at::BFloat16));
  auto norms = at::empty_strided(g.sizes(), g.strides(), g.options().dtype(AccType));


  const int ndims = v.dim();

  if(dim == 0)
  {
    // Find logical size of each flattened slowest-dim row
    int rowSize = 1;
    for(int i = ndims - 1; i > 0; i--)
      rowSize *= v.size(i);

    cudaStream_t stream = at::cuda::getCurrentCUDAStream();

    printf("____________test_1____________\n");
    AT_DISPATCH_FLOATING_TYPES_AND2(
        kHalf,
        kBFloat16,
        v.scalar_type(), "weight_norm_fwd_first_dim_kernel", [&] {
          using accscalar_t = acc_type<scalar_t, true>;
          using opmath_t = at::opmath_type<scalar_t>;

          std::cout << "test code section norms.mutable_data_ptr in!" << std::endl;
          std::cout << typeid(scalar_t).name() << std::endl; // check scalar_t type
          std::cout << typeid(accscalar_t).name() << std::endl; // check accscalar_t type
          std::cout << typeid(opmath_t).name() << std::endl; // check opmath_t type

          std::cout << norms.dtype() << std::endl; // check norms dtype

          scalar_t* x = norms.mutable_data_ptr<scalar_t>();
          accscalar_t* y = (accscalar_t*) x;

          std::cout << typeid(y).name() << std::endl; // check TypeCasted obj dtype
          std::cout << "test code section norms.mutable_data_ptr out!" << std::endl;

          weight_norm_fwd_first_dim_kernel<scalar_t, accscalar_t>
              <<<v.size(0), BLOCK, BLOCK * sizeof(accscalar_t), stream>>>(
                  w.mutable_data_ptr<scalar_t>(),
                  y, // tmp, it just for testing
                  v.const_data_ptr<scalar_t>(),
                  g.const_data_ptr<scalar_t>(),
                  rowSize);
          C10_CUDA_KERNEL_LAUNCH_CHECK();
        });
  }
  else if(dim == ndims - 1)
  {
    // Precompute slower_dims_size and fast_dim_size
    int slower_dims_size = 1;
    for(int i = 0; i < ndims - 1; i++)
      slower_dims_size *= v.size(i);

    int fast_dim_size = v.size(ndims-1);

    cudaStream_t stream = at::cuda::getCurrentCUDAStream();

    printf("____________test_2____________\n");
    AT_DISPATCH_FLOATING_TYPES_AND2(
        kHalf,
        kBFloat16,
        v.scalar_type(),
        "weight_norm_fwd_last_dim_kernel",
        [&] {
          using accscalar_t = acc_type<scalar_t, true>;
          std::cout << typeid(accscalar_t).name() << std::endl;

          weight_norm_fwd_last_dim_kernel<scalar_t, accscalar_t>
              <<<(fast_dim_size + TILE_W - 1) / TILE_W,
                 dim3(TILE_W, TILE_H),
                 (TILE_W * TILE_H + TILE_W) * sizeof(accscalar_t),
                 stream>>>(
                  w.mutable_data_ptr<scalar_t>(),
                  norms.mutable_data_ptr<accscalar_t>(),
                  v.const_data_ptr<scalar_t>(),
                  g.const_data_ptr<scalar_t>(),
                  fast_dim_size,
                  slower_dims_size);
          C10_CUDA_KERNEL_LAUNCH_CHECK();
        });
  }

  // The kernel execution is asynchronous, so this will only catch errors on the kernel launch,
  // not the kernel's execution.  Errors in kernel execution aren't guaranteed to be caught
  // until a later error check on a synchronizing CUDA call.  Unfortunately, without manually
  // synchronizing here, the foregoing is the best we can do.

  return std::tuple<Tensor, Tensor>{w, norms};
}
  }
```

**Review by jp1924 — COMMENTED** [review]
When weight_norm is ported to parametrizations, it can result in overloading, so is there a need to check `self.assertRaisesRegex`?

**Comment by jp1924** [comment]
@mikaylagawarecki 
I don't think I can do this PR. Seems like it should be done by someone else, or someone who is good at C++.
Thanks for the help though!

**Pull request closed without merging**